libraries-3/src/test/java/com/baeldung/jnats/NatsClientLiveTest.java

```diff
@@ -1,115 +1,336 @@
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
+            List<Message> messages = new ArrayList<>();
 
-        Message message = fooSubscription.nextMessage(200);
-        assertNull("Got message!", message);
+            Message message = qSub1.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
 
+            message = qSub2.nextMessage(TIMEOUT_MILLIS);
+            if (message != null) {
+                messages.add(message);
+            }
 
-        SyncSubscription barSubscription = client.subscribeSync("foo.>");
+            assertEquals(1, messages.size());
 
-        client.publishMessage("foo.bar.plop", "bar.foo", "hello there");
-
-        message = barSubscription.nextMessage(200);
-        assertNotNull("No message!", message);
-        assertEquals("hello there", new String(message.getData()));
+            qSub1.unsubscribe();
+            qSub2.unsubscribe();
+        }
+    }
 
+    @Test
+    public void whenSubscribingAsynchronouslyWithoutQueueGroups_thenEachMessageHandlerShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            List<Message> messages1 = new ArrayList<>();
+            List<Message> messages2 = new ArrayList<>();
+
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            Subscription subscription1 = dispatcher.subscribe("mySubject", messages1::add);
+            Subscription subscription2 = dispatcher.subscribe("mySubject", messages2::add);
+
+            natsConnection.publish("mySubject", "data1".getBytes());
+            natsConnection.publish("mySubject", "data2".getBytes());
+
+            // Simulate real world time for messages to propagate.
+            Awaitility.await()
+                .atMost(WAIT_AT_MOST_MILLIS, TimeUnit.MILLISECONDS)
+                .pollDelay(POLL_DELAY_MILLIS, TimeUnit.MILLISECONDS)
+                .until(() -> messages1.size() + messages2.size() == 4);
+
+            assertEquals(2, messages1.size());
+            assertEquals(2, messages2.size());
+
+            dispatcher.unsubscribe(subscription1);
+            dispatcher.unsubscribe(subscription2);
+        }
+    }
 
+    @Test
+    public void whenSubscribingAsynchronouslyWithQueueGroups_thenOnlyOneMessageHandlerInTheGroupShouldReceiveEachMessage() throws Exception {
+        try (Connection natsConnection = createConnection()) {
+            List<Message> messages = new ArrayList<>();
+
+            Dispatcher dispatcher = natsConnection.createDispatcher();
+            Subscription qSub1 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+            Subscription qSub2 = dispatcher.subscribe("mySubject", "myQueue", messages::add);
+
+            natsConnection.publish("mySubject", "data".getBytes());
+
+            // Simulate real world time for messages to propagate.
```

Comment:
We aren't really simulating anything here, we are literally waiting for the message to arrive
```suggestion
            // wait for messages to propagate.
```

Reply:
fixed